web_pwa_cache/static/src/js/worker/bus.js

```diff
@@ -0,0 +1,112 @@
+/* Copyright 2020 Tecnativa - Alexandre D. Díaz
+ * License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). */
+
+odoo.define("web_pwa_cache.PWA.bus", function(require) {
+    "use strict";
+
+    const BroadcastMixin = require("web_pwa_cache.BroadcastMixin");
+    const PWA = require("web_pwa_oca.PWA");
+    require("web_pwa_cache.PWA");
+
+    /**
+     * This class is used to communicate with the user page.
+     */
+    PWA.include(BroadcastMixin);
+    PWA.include({
+        /**
+         * @override
+         */
+        init: function() {
+            this.init_broadcast("pwa-sw-messages", "pwa-page-messages");
+            this._super.apply(this, arguments);
+        },
+
+        /**
+         * @override
+         */
+        _onCacheNotFound: function(request, err) {
+            this._super.apply(this, arguments);
+            const isOffline =
+                (this._managers.config && this._managers.config.isOfflineMode()) ||
+                false;
+            if (isOffline) {
+                const request_url = new URL(request.url);
+                this.postBroadcastMessage({
+                    type: "PWA_CACHE_FAIL",
+                    error: err,
+                    url: request_url.pathname,
+                });
+            }
+        },
+
+        /**
+         * @param {String} type
+         * @param {Object} data
+         * @returns {Promise}
+         */
+        onProcessBusMessage: function(type, data) {
+            if (type === "START_PREFETCH") {
+                if (this.isActivated()) {
+                    return this._doPrefetchDataPost();
+                }
+                return Promise.resolve();
+            }
+
+            const tasks = [];
+            for (const manager_name in this._managers) {
+                tasks.push(this._managers[manager_name].onProcessMessage(type, data));
+            }
+            return Promise.all(tasks);
+        },
+
+        /**
+         * Initialize the prefetching process.
+         * Try to sync. records first.
+         *
+         * @returns {Promise}
+         */
+        _doPrefetchDataPost: function() {
+            if (this._prefetch_promise) {
+                return this._prefetch_promise;
+            }
+
+            this._prefetch_promise = new Promise(async (resolve, reject) => {
+                this._prefetch_running = true;
+                // Try sync records first
+                try {
+                    await this._managers.sync.run();
```

Comment:
Here I'm getting `TypeError: Cannot read properties of undefined (reading 'run')`. It seems that `sync` is not there in `_managers`, only `config`, and `activateWorker`, the place where it's supposed to be filled, does not seem to be called. As a result the UserMenu is not shown.

Reply:
Hi! Please, can you give me more info or/and steps to reproduce?

For example...
The user has the "Enable PWA Cache" group?
Browser?
This is present after pwa installation? before?

Thanks.

Reply:
Actually it was because my `admin` user didnt have that `Enable PWA Cache` group enabled - I didnt know I was supposed to do that.

So I guess that solves my issue just maybe leads to some minor comments such as

- Add [Installation note](https://github.com/Tecnativa/web/tree/13.0-add-web_pwa_cache/web_pwa_cache/readme) mentioning that group, also that Chrome is required and Firefox wont work (and why)
- Add the group also to user form view (had to go to Groups menu to fill it)
- Better handling of this error and the other one I commented separately - these occur when you have the PWA installed on Chrome 93 on Ubuntu, but your user doesnt have the group.

Reply:
Great work by the way :+1:

Reply:
Thanks! i will check the issues...

P.S: 'admin' user is not the best option to use with pwa cache... you will get all possible records.

Reply:
Updated Readme's ... i can't reproduce the issue with "sync" :/